R/aggregate.R

```diff
@@ -1,222 +1,123 @@
-##' @return `numeric()` vector of length `ncol(x)` with robust
-##'     summarised values.
-##' 
-##' @author Adriaan Sticker, Sebastian Gibb and Laurent Gatto
 ##'
-##' @family Quantitative feature aggregation
-##' 
-##' @export
+##' These functions take a matrix of quantitative features `x` and
+##' aggregate the features (rows) according to either a vector (or
+##' factor) `INDEX` or an adjacency matrix `MAT`. The aggregation
+##' method is defined by function `FUN`.
 ##'
-##' @importFrom MASS rlm
-##' @importFrom stats model.matrix .lm.fit
-##' 
-##' @examples
-##' x <- matrix(rnorm(30), nrow = 3)
-##' colnames(x) <- letters[1:10]
-##' rownames(x) <- LETTERS[1:3]
-##' robustSummary(x)
-robustSummary <- function(x, ...) {
-    if (is.null(colnames(x)))
-        stop("colnames must not be empty.")
-    if (is.null(rownames(x)))
-        stop("rownames must not be empty.")
-
-    ## If there is only one 1 peptide for all samples return
-    ## expression of that peptide
-    if (nrow(x) == 1L) return(x)
-
-    ## remove missing values
-    p <- !is.na(x)
-    expression <- x[p] ## expression becomes a vector
-    sample <- rep(colnames(x), each = nrow(x))[p]
-    feature <- rep(rownames(x), times = ncol(x))[p]
-
-    ## model.matrix breaks on factors with 1 level so make vector of
-    ## ones (intercept).
-    if (length(unique(sample)) == 1L) sample <- rep(1, length(sample))
-
-    ## Sum contrast on peptide level so sample effect will be mean
-    ## over all peptides instead of reference level.
-    X <- model.matrix(~ -1 + sample + feature,
-                      contrasts.arg = list(feature = 'contr.sum'))
-    ## MASS::rlm breaks on singulare values.
-    ## - Check with base lm if singular values are present.
-    ## - If so, these coefficients will be zero, remove this collumn
-    ##   from model matrix
-    ## - Rinse and repeat on reduced modelmatrix till no singular
-    ##   values are present
-    repeat {
-        fit <- .lm.fit(X, expression)
-        id <- fit$coefficients != 0
-        X <- X[ , id, drop = FALSE]
-        if (all(id)) break
-    }
-    ## Last step is always rlm: calculate estimated effects as
-    ## summarised values
-    fit <- rlm(X, expression, ...)
-
-    sampleid <- seq_along(unique(sample))
-    ## This will be needed for NUSE-type of quality control, but will
-    ## need to check for missing data as below.
-    ## se <- unique(summary(fit)$coefficients[sampleid, 'Std. Error'])
-
-    ## Take the sample coefficients ( = summarised expression values)
-    coef  <-  fit$coefficients[sampleid]
-    ## Sort the sample coefficients in the same way as the samplenames
-    ## of expression matrix. Puts NA for the samples without any
-    ## expression value
-    res <- coef[paste0('sample', colnames(x))]
-    names(res) <- colnames(x)
-    res
-}
-
-##' @title Return the Median Polish (Robust Twoway Decomposition) of a matrix
+##' Adjacency matrices are an elegant way to explicitly encode for
```

Comment:
Just out of curiosity: what's the advantage of a `matrix`-like structure over a `list` of `integer` indices? I had the feeling that sparse matrices can be sometimes quite slow.

Reply:
For a real-life scenario of an adjacency matrix, see the [makeAdjacencyMatrix()](https://rformassspectrometry.github.io/PSMatch/reference/adjacencyMatrix.html) manual page. I compared the implementation of that function on an mzID file using a regular and a sparse matrix, and the latter was smaller (as expected) but the code was substantially faster (at least with my simple implementations). I think the difference can be explained by the way the matrices are created:

```r
> m <- matrix(0, nrow = 3, ncol = 2)
> m[1, 1] <- m[2, 1]  <- m[3, 2]  <- 1 ## actually a for loop 
> 
> m
     [,1] [,2]
[1,]    1    0
[2,]    1    0
[3,]    0    1
> library(Matrix)
> sparseMatrix(i = c(1, 2, 3), j = c(1, 1, 2), x = 1)
3 x 2 sparse Matrix of class "dgCMatrix"
        
[1,] 1 .
[2,] 1 .
[3,] . 1
```
I also tested a couple of relevant cross products with these adjacency matrices, and sparse matrices were efficient.

Reply:
thanks for the clarification